app/main.py

```diff
@@ -1,214 +1,39 @@
-            }
-        }
-    }
-}
-
-
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
-
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
-
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
-
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
-
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
-
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
-
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
-
-    # arthur
-    arthur = knightsConfig["arthur"]
-
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
-
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
-
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
-
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
+from app.battle_preparations.prepare_to_battle import preparations
 
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
 
-    # mordred
-    mordred = knightsConfig["mordred"]
+def battle(knights: dict) -> dict:
+    knights_instances = preparations(knights)
+    lancelot, arthur, mordred, red_knight = knights_instances
 
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
+    lancelots_attack = lancelot.attack(mordred.hp, mordred.protection)
 
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
+    if lancelots_attack <= 0:
+        mordred.hp = 0
+    else:
+        mordred.hp = lancelots_attack
 
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
+    mordreds_attack = mordred.attack(lancelot.hp, lancelot.protection)
 
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
+    if mordreds_attack <= 0:
+        lancelot.hp = 0
+    else:
+        lancelot.hp = mordreds_attack
 
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
+    arthur_attack = arthur.attack(red_knight.hp, red_knight.protection)
+    if arthur_attack <= 0:
+        red_knight.hp = 0
+    else:
+        red_knight.hp = arthur_attack
```

Comment:
You could separate this logic and place it in the attack so that there would be no need for code duplication on lines 29 up to 32 and similar code above.